File: scripts/test_grader.py
#!/usr/bin/env python3
"""Shared-fixture test for the Python grader/annotation parser (mirror of scripts/test_grader.mjs)."""
import json, sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parent))
import build
fx = json.loads((Path(__file__).resolve().parent.parent / "tests" / "grader-fixtures.json").read_text())
failed = 0
for t in fx["annotations"]:
    got = build.parse_annotations(t["text"], t["ruleId"])
    ok = got == t["expect"]; failed += not ok
    print(f"  {'ok  ' if ok else 'FAIL'} annotations: {t['name']}" + ("" if ok else f" got={got}"))
for t in fx["grades"]:
    g = build.grade(t["result"], t["expected"], t["wasm"])
    got = {k: g[k] for k in t["expect"]}
    ok = got == t["expect"]; failed += not ok
    print(f"  {'ok  ' if ok else 'FAIL'} grade: {t['name']}" + ("" if ok else f" got={got}"))
print("python grader:", "all fixtures pass" if not failed else f"{failed} failure(s)")
sys.exit(1 if failed else 0)

File: scripts/test_grader.mjs
#!/usr/bin/env node
// Shared-fixture test for the JS grader/annotation parser (mirror of scripts/test_grader.py).
import { readFileSync } from 'node:fs';
import { grade } from '../docs/js/grader.js';
import { parseAnnotations } from '../docs/js/annotations.js';
const fx = JSON.parse(readFileSync(new URL('../tests/grader-fixtures.json', import.meta.url), 'utf8'));
let failed = 0;
const same = (a, b) => JSON.stringify(a) === JSON.stringify(b);
for (const t of fx.annotations) {
  const got = parseAnnotations(t.text, t.ruleId);
  const ok = same(got, t.expect); failed += !ok;
  console.log(`  ${ok ? 'ok  ' : 'FAIL'} annotations: ${t.name}${ok ? '' : ' got=' + JSON.stringify(got)}`);
}
for (const t of fx.grades) {
  const g = grade(t.result, t.expected, { wasm: t.wasm });
  const got = Object.fromEntries(Object.keys(t.expect).map((k) => [k, g[k]]));
  const ok = same(got, t.expect); failed += !ok;
  console.log(`  ${ok ? 'ok  ' : 'FAIL'} grade: ${t.name}${ok ? '' : ' got=' + JSON.stringify(got)}`);
}
console.log('js grader:', failed ? `${failed} failure(s)` : 'all fixtures pass');
process.exit(failed ? 1 : 0);
